src/main/haskell/matching-logic/src/Kore/MatchingLogic/Signature/Simple.hs

```diff
@@ -0,0 +1,120 @@
+  ,ValidatedSignature,fromValidated,validate,reifySignature
+  ,findLabel,findSort,recognizeSignatureSyntax) where
+import           Data.Char
+import           Data.Coerce
+import           Data.Functor.Foldable
+import           Data.Map.Strict           (Map)
+import qualified Data.Map.Strict           as Map
+import           Data.Proxy
+import           Data.Set                  (Set)
+import qualified Data.Set                  as Set
+import           Data.Text                 (Text)
+import qualified Data.Text                 as Text
+import           Data.Text.Prettyprint.Doc
+
+import           Data.Reflection
+
+import           Kore.MatchingLogic.AST
+
+data SignatureInfo =
+  MkSignatureInfo
+  { sorts  :: Set Text
+  , labels :: Map Text (Text,[Text])
+  }
+  deriving Show
+
+newtype ValidatedSignature =
+  MkValidatedSignature SignatureInfo
+instance Show ValidatedSignature where
+  show (MkValidatedSignature sig) = show sig
+
+fromValidated :: ValidatedSignature -> SignatureInfo
+fromValidated (MkValidatedSignature info) = info
+
+data SimpleSignature s
+
+isValid :: SignatureInfo -> Bool
+isValid sigInfo =
+  all (\(r,args) -> sortOk r && all sortOk args) (labels sigInfo)
+ where sortOk s = Set.member s (sorts sigInfo)
+
+validate :: SignatureInfo -> Maybe ValidatedSignature
+validate sig = if isValid sig then Just (MkValidatedSignature sig) else Nothing
+
+findSort :: forall s . (Reifies s ValidatedSignature) => Text -> Maybe (Sort (SimpleSignature s))
+findSort sortName =
+  case Set.lookupGE sortName (sorts sig) of
+    Just sort' | sort' == sortName -> Just (SimpleSort sort')
+    _          -> Nothing
+ where sig = fromValidated (reflect @s Proxy)
+
+findLabel :: forall s . (Reifies s ValidatedSignature) => Text -> Maybe (Label (SimpleSignature s))
+findLabel labelName =
+  case Map.lookupGE labelName (labels sig) of
+    Just (label',_) | label' == labelName -> Just (SimpleLabel label')
+    _               -> Nothing
+ where sig = fromValidated (reflect @s Proxy)
+
+instance Show (Label (SimpleSignature s)) where
+  show (SimpleLabel l) = show l
+instance Show (Sort (SimpleSignature s)) where
+  show (SimpleSort l) = show l
+
+instance Pretty (Label (SimpleSignature s)) where
+  pretty (SimpleLabel l)
+    | isAlpha (Text.head l) && Text.all isAlphaNum l = pretty l
+    | otherwise = pretty (show l)
+instance Pretty (Sort (SimpleSignature s)) where
+  pretty (SimpleSort l)
+    | isAlpha (Text.head l) && Text.all isAlphaNum l = pretty l
+    | otherwise = pretty (show l)
+
+instance (Reifies s ValidatedSignature) => IsSignature (SimpleSignature s) where
+  newtype Label (SimpleSignature s) = SimpleLabel Text
+  newtype Sort (SimpleSignature s) = SimpleSort Text
+  labelSignature (SimpleLabel name) =
+    case Map.lookup name (labels sig) of
+      Just labelSignature -> coerce labelSignature
+      Nothing -> error $ "Encapsulation failure, invalid label "++show name++" found in a reflected signature"
```

Comment:
I think that Main.hs has a user interface, and this is used there, right? If so, how much will that UI be used? If it is/will be used more than for demo purposes, why is it reasonable to stop the program for a typo?

Regardless of whether that UI exists/matters:
Why would you rather have runtime errors instead of compile-time errors? The latter means that an entire class of bugs simply can't happen.

Reply:
That line should be impossible. I've just put a more descriptive error in case the impossible happens.

Reply:
This is a very crude UI, but the plan is that eventually we will have a nice interactive prover.